### PR TITLE
Show an error if CIA contend is encrypted

### PR DESCRIPTION
**Pull request by B3n30**
We already show an error if the cia is encrypted but not if the cia itself isn't encrypted but it's content is.
This lead to a lot of confusion where users thought they successfully installed the cia, but then nothing appeared in the game list. Now the encryption error is also shown if the cia contained any executable that Citra isn't able to decrypt (e.g. due to missing aes keys).

In addition any executable that is encrypted and Citra can't decrypt is still shown in the game list (if executables without icons are visible). Such executables will still be shown in an ugly way (no proper title, icon, region etc), since we can't access the SMDH.

As a last change in that PR I removed a debugging line we added that showed the path of each NCCH file, since it was only added to debug an issue where loading encrypted NCCHs caused a crash.

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/5130)

### CODE REVIEW

**Comment by B3n30** [comment]
This should fix #5065 and #4117

**Comment by FearlessTobi** [comment]
The braces should be placed on the same line as the if statements.

File: src/core/file_sys/ncch_container.cpp [hunk 1]
```diff
@@ -133,8 +133,37 @@ Loader::ResultStatus NCCHContainer::OpenFile(const std::string& filepath, u32 nc
     return Loader::ResultStatus::Success;
 }
 
+Loader::ResultStatus NCCHContainer::LoadHeader() {
+    if (has_header)
+        return Loader::ResultStatus::Success;
+    if (!file.IsOpen()) {
+
```
Comment (FearlessTobi) [inline src/core/file_sys/ncch_container.cpp]: Remove this empty line.

File: src/core/file_sys/ncch_container.cpp [hunk 2]
```diff
@@ -133,8 +133,37 @@ Loader::ResultStatus NCCHContainer::OpenFile(const std::string& filepath, u32 nc
     return Loader::ResultStatus::Success;
 }
 
+Loader::ResultStatus NCCHContainer::LoadHeader() {
+    if (has_header)
```
Comment (FearlessTobi) [inline src/core/file_sys/ncch_container.cpp]: I think we are following the convention that all if statements should be braced. Dito below.

File: src/core/file_sys/ncch_container.cpp (hunk 1 above)
Comment (vitor-k) [inline src/core/file_sys/ncch_container.cpp]: Unnecessary new line

File: src/core/hle/service/am/am.cpp [hunk 3]
```diff
@@ -33,6 +34,16 @@
 #include "core/loader/loader.h"
 #include "core/loader/smdh.h"
 
+namespace {
+bool HasSupportedFileExtension(std::string path) {
+    static const std::array<std::string, 7> extensions = {{".3ds", ".3dsx", ".elf", ".axf", ".cci",
+                                                           ".cxi"
```
Comment (vitor-k) [inline src/core/hle/service/am/am.cpp]: Is the lack of a comma here on purpose?

File: src/citra_qt/game_list_worker.cpp [hunk 4]
```diff
@@ -46,13 +46,15 @@ void GameListWorker::AddFstEntriesToGameList(const std::string& dir_path, unsign
...
             std::unique_ptr<Loader::AppLoader> loader = Loader::GetLoader(physical_name);
-            if (!loader)
+            if (!loader) {
                 return true;
+            }
 
             bool executable = false;
-            loader->IsExecutable(executable);
-            if (!executable)
+            auto res = loader->IsExecutable(executable);
```
Comment (FearlessTobi) [inline src/citra_qt/game_list_worker.cpp]: Can this be "const"?

File: src/core/hle/service/am/am.cpp [hunk 5]
```diff
@@ -373,6 +384,32 @@ InstallStatus InstallCIA(const std::string& path,
...
+            const std::string physical_name = directory + DIR_SEP + virtual_name;
+            const bool is_dir = FileUtil::IsDirectory(physical_name);
+            if (!is_dir && HasSupportedFileExtension(physical_name)) {
+                std::unique_ptr<Loader::AppLoader> loader = Loader::GetLoader(physical_name);
+                if (!loader) {
+                    return true;
+                }
+
+                bool executable = false;
+                auto res = loader->IsExecutable(executable);
```
Comment (FearlessTobi) [inline src/core/hle/service/am/am.cpp]: Can this be "const"?

File: src/core/hle/service/am/am.cpp (hunk 3 above)
Comment (B3n30) [reply src/core/hle/service/am/am.cpp]: no. clang-format seems to messed this up too.
Now it should be fine

**Comment by B3n30** [comment]
Since some users requested it, I added a log that shows which key is missing when a Ticket gets decrypted during cia installation

File: src/core/hle/service/am/am.cpp [hunk 6]
```diff
@@ -33,6 +34,15 @@
...
 #include "core/loader/smdh.h"
 
+namespace {
+bool HasSupportedFileExtension(std::string path) {
+    static const std::array<std::string, 7> extensions = {
+        {".3ds", ".3dsx", ".elf", ".axf", ".cci", ".cxi", ".app"}};
+    const auto file_ext = FileUtil::GetExtensionFromFilename(path);
+    return std::find(extensions.begin(), extensions.end(), file_ext) != extensions.end();
+}
+} // namespace
```
Comment (zhaowenlan1779) [inline src/core/hle/service/am/am.cpp]: I personally don't see a reason for this hardcoded list of extensions, why can't we just try to load everything in the folder?

File: src/core/file_sys/ncch_container.cpp [hunk 7]
```diff
@@ -697,7 +728,7 @@ Loader::ResultStatus NCCHContainer::ReadOverrideRomFS(std::shared_ptr<RomFSReade
 }
 
 Loader::ResultStatus NCCHContainer::ReadProgramId(u64_le& program_id) {
-    Loader::ResultStatus result = Load();
+    Loader::ResultStatus result = LoadHeader();
```
Comment (zhaowenlan1779) [inline src/core/file_sys/ncch_container.cpp]: Why is this change necessary?

File: src/core/hle/service/am/am.cpp (hunk 6 above)
Comment (B3n30) [reply src/core/hle/service/am/am.cpp]: This was just to don't do any unnecessary stuff (on big game lists this could really be costly)

File: src/core/file_sys/ncch_container.cpp (hunk 7 above)
Comment (B3n30) [reply src/core/file_sys/ncch_container.cpp]: This ensure that we get a ProgramID even if the NCCH is encrypted and the keys aren't available. It probably still requires to change the ui to show the programId instead of the filename in the game list though

File: src/core/file_sys/ncch_container.cpp [hunk 8]
```diff
@@ -133,8 +133,39 @@ Loader::ResultStatus NCCHContainer::OpenFile(const std::string& filepath, u32 nc
...
+    }
+
+    // Verify we are loading the correct file type...
+    if (Loader::MakeMagic('N', 'C', 'C', 'H') != ncch_header.magic) {
+        return Loader::ResultStatus::ErrorInvalidFormat;
+    }
+
+    has_header = true;
+    return Loader::ResultStatus::Success;
+}
```
Comment (zhaowenlan1779) [inline src/core/file_sys/ncch_container.cpp]: Should probably convert the first part of the `Load()` to `LoadHeader()` then

File: src/core/hle/service/am/am.cpp [hunk 9]
```diff
@@ -373,6 +383,32 @@ InstallStatus InstallCIA(const std::string& path,
...
+                const auto res = loader->IsExecutable(executable);
+                if (res == Loader::ResultStatus::ErrorEncrypted) {
+                    return false;
+                }
+                return true;
+            } else {
+                return FileUtil::ForeachDirectoryEntry(nullptr, physical_name, callback);
+            }
+        };
+        if (!FileUtil::ForeachDirectoryEntry(nullptr, path, callback)) {
```
Comment (zhaowenlan1779) [inline src/core/hle/service/am/am.cpp]: Is this correct? won't `path` point to the CIA file (instead of the installed directory)?

File: src/core/hle/service/am/am.cpp (hunk 6 above)
Comment (zhaowenlan1779) [reply src/core/hle/service/am/am.cpp]: This is only executed once when a CIA is installed so I don't think it's going to influence the game list performance

File: src/core/file_sys/ncch_container.cpp (hunk 8 above)
Comment (B3n30) [reply src/core/file_sys/ncch_container.cpp]: Thats not that easy because it seems the remaining part of Load() needs the file to be at the position after the header

File: src/core/hle/service/am/am.cpp (hunk 9 above)
Comment (B3n30) [reply src/core/hle/service/am/am.cpp]: Oh yeah, thx for spotting this
